src/dmd/dtoh.d

```diff
@@ -0,0 +1,1256 @@
+    buf.writestring("#define _d_dchar unsigned\n");
+    buf.writestring("\n");
+    buf.writestring("#define _d_null NULL\n");
+    buf.writestring("\n");
+    buf.writestring("struct AA;\n");
+    buf.writestring("\n");
+
+    OutBuffer check;
+    check.writestring(`
+#if OFFSETS
+
+template <class T>
+size_t getSlotNumber(int dummy, ...)
+{
+    T c;
+    va_list ap;
+    va_start(ap, dummy);
+    void *f = va_arg(ap, void*);
+    for (size_t i = 0; ; i++)
+    {
+        if ( (*(void***)&c)[i] == f)
+            return i;
+    }
+    va_end(ap);
+}
+
+void testOffsets()
+{
+    `);
+
+    OutBuffer done;
+    OutBuffer decl;
+    scope v = new ToCppBuffer!ASTCodegen(&check, buf, &done, &decl);
+    foreach (m; *ms)
+    {
+        buf.printf("// Parsing module %s\n", m.toPrettyChars());
+        m.accept(v);
+    }
+    buf.write(&done);
+    buf.write(&decl);
+
+    check.writestring(`
+}
+#endif
+    `);
+
+    debug buf.write(&check);
+}
+
+void main()
+{
+    import std.stdio;
+    import std.algorithm.sorting : sort;
+    import std.array : array;
+    import std.file : readText;
+    import std.path : baseName, buildPath, dirName;
+    import std.string : toStringz;
+
+    import dmd.id;
+    import dmd.parse;
+    import dmd.dsymbolsem;
+    import dmd.semantic2;
+    import dmd.semantic3;
+    import dmd.builtin : builtin_init;
+    import dmd.dmodule : Module;
+    import dmd.cond : VersionCondition;
+    import dmd.expression : Expression;
+    import dmd.objc : Objc;
+    import dmd.target : Target;
+
+    import core.memory;
+
+    GC.disable();
+
+    global._init();
+    global.params.isLinux = true;
+    global.params.is64bit = (size_t.sizeof == 8);
+    global.params.isLP64 = global.params.is64bit;
+
+    global.path = new Strings();
+    global.path.push(__FILE_FULL_PATH__.dirName.buildPath("../../../druntime/src/").toStringz());
+
+    global.filePath = new Strings();
+    global.params.fileImppath = global.filePath;
+    global.filePath.push(__FILE_FULL_PATH__.dirName.buildPath("../../").toStringz());
+    global.filePath.push(__FILE_FULL_PATH__.dirName.buildPath("../../res/").toStringz());
+
+    ASTCodegen.Type._init();
+    Id.initialize();
+    Module._init();
+    Target._init();
+    Expression._init();
+    Objc._init();
+    builtin_init();
+
+    VersionCondition.addPredefinedGlobalIdent("all");
+    VersionCondition.addPredefinedGlobalIdent("DigitalMars");
+    VersionCondition.addPredefinedGlobalIdent("Posix");
+    VersionCondition.addPredefinedGlobalIdent("linux");
+    VersionCondition.addPredefinedGlobalIdent("CRuntime_Glibc");
+    VersionCondition.addPredefinedGlobalIdent("NoBackend");
+    VersionCondition.addPredefinedGlobalIdent("NoMain");
+    if (global.params.is64bit)
+        VersionCondition.addPredefinedGlobalIdent("X86_64");
+    else
+        VersionCondition.addPredefinedGlobalIdent("X86");
+    if (global.params.isLP64)
+        VersionCondition.addPredefinedGlobalIdent("D_LP64");
```

Comment:
The frontend module (available via the dub package) is intended to avoid exactly this boilerplate code.
See https://github.com/dlang/dmd/blob/master/src/dmd/frontend.d and e.g. https://github.com/dlang/dmd/blob/master/test/dub_package/frontend.d